File: frontend/src/utils/contentFilters.ts
/** Process vs final answer filtering — keep rules separate. */

export function hasRichMarkdownReport(text: string): boolean {
  const t = (text || '').trim()
  if (t.length < 120) return false
  const headers = (t.match(/^#{1,6}\s+\S/gm) || []).length
  const hasTable = t.includes('|') && t.includes('---')
  if (headers >= 2 && hasTable) return true
  if (headers >= 1 && hasTable && t.length >= 200) return true
  return false
}

export function isProcessTemplateDump(text: string): boolean {
  const t = (text || '').trim()
  if (!t) return true
  const compact = t.replace(/\s/g, '')
  if (compact.includes('##分析过程') || compact.includes('###步骤')) return true
  if (
    ['-类型:', '- 类型:', '等待执行', '等待返回', '状态:进行中', '状态:等待', '-状态:进行中', '-状态:等待'].some(
      (token) => t.includes(token),
    )
  ) {
    return true
  }
  if (
    (t.match(/✅/g) || []).length >= 2 &&
    !['QED', '评分', 'pocket', 'PDB', '|', '对接', 'hERG', 'BBB'].some((m) => t.includes(m))
  ) {
    return true
  }
  return false
}

export function isSynthesizedStepDumpFinal(text: string): boolean {
  const t = (text || '').trim()
  if (!t) return false
  const fieldHits = (t.match(/^-\s*(?:输入摘要|结果摘要|详情):/gm) || []).length
  if (fieldHits >= 2) return true
  if (
    t.includes('逆合成路线已生成') &&
    (t.includes('未能找到') || t.includes('no synthesis routes') || t.includes('MCP 工具未能'))
  ) {
    return true
  }
  if (t.startsWith('**3D构象生成**') && t.includes('逆合成路线已生成')) return true
  return false
}

/** Final answer unusable → fall back to step synthesis only. */
export function isFinalAnswerUnusable(text: string): boolean {
  const t = (text || '').trim()
  if (!t) return true
  if (isSynthesizedStepDumpFinal(t)) return true
  if (hasRichMarkdownReport(t)) return false
  if (isProcessTemplateDump(t)) return true
  if (t.includes('"pocket_id":') && !t.includes('|') && t.length < 200) return true
  if (t.includes('"molecules":') && !t.includes('|') && t.length < 200) return true
  return false
}

const EMOJI_BETWEEN = '(?:[\\u{1F300}-\\u{1FAFF}]|[\\u{2600}-\\u{27BF}])\\s*'

function stripOrphanEmojiLines(text: string): string {
  return (text || '')
    .split('\n')
    .filter((ln) => !/^\s*(?:[\u{1F300}-\u{1FAFF}]|[\u{2600}-\u{27BF}])+\s*$/u.test(ln))
    .join('\n')
}

export function stripStructuredJsonAppendix(text: string): string {
  let t = (text || '').trim()
  if (!t) return ''
  t = t
    .replace(
      new RegExp(
        `\\n*(?:#{1,3}\\s*)?${EMOJI_BETWEEN}(?:\\*{1,2})?完整结构化数据:?(?:\\*{1,2})?\\s*\\n*\`\`\`json\\s*[\\s\\S]*?\`\`\``,
        'giu',
      ),
      '',
    )
    .trim()
  t = t.replace(/\n*```json\s*\{[\s\S]*?"session_id"[\s\S]*?```\s*$/i, '').trim()
  t = t
    .replace(new RegExp(`\\n*#{1,3}\\s*${EMOJI_BETWEEN}完整结构化数据\\s*$`, 'giu'), '')
    .trim()
  return stripOrphanEmojiLines(t)
}

export function sanitizeFinalAnswerText(text: string): string {
  let t = (text || '').trim()
  if (!t) return ''
  const procIdx = t.search(/##\s*分析过程/)
  if (procIdx >= 0) t = t.slice(0, procIdx).trim()
  const noiseIdx = t.search(/\n⚠️/)
  if (noiseIdx >= 0) t = t.slice(0, noiseIdx).trim()
  const execIdx = t.indexOf('\n⚠️ 🛠️ Exec failed:')
  if (execIdx >= 0) t = t.slice(0, execIdx).trim()
  return stripStructuredJsonAppendix(t)
}

/** @deprecated Use isFinalAnswerUnusable */
export function isLowQualityFinalAnswer(text: string): boolean {
  return isFinalAnswerUnusable(text)
}

const TRAILING_FINAL_START =
  /(?:✅\s*\*\*第\s*\d+\s*步|✅\s*\*\*步骤\s*\d+|(?:^|\n)##\s*结果摘要|(?:^|\n)###\s*🔬|(?:^|\n)#\s+🔬|(?:^|\n)###\s*🫁)/m

function looksLikeLiveProcessChunk(text: string): boolean {
  const t = (text || '').trim()
  if (!t) return true
  if (t.includes('### 步骤') && t.includes('- 类型:') && t.includes('- 状态:')) return true
  if (t.includes('- 工具数:') && t.includes('### 步骤')) return true
  return false
}

/** Model prose after embedded process when ## 最终回答 is missing. */
export function extractTrailingModelFinal(raw: string): string {
  const text = (raw || '').trim()
  if (!text) return ''

  const candidates: string[] = []
  const re = new RegExp(TRAILING_FINAL_START.source, TRAILING_FINAL_START.flags + 'g')
  let m: RegExpExecArray | null
  while ((m = re.exec(text)) !== null) {
    const tail = sanitizeFinalAnswerText(text.slice(m.index))
    if (tail && !isProcessTemplateDump(tail) && !isFinalAnswerUnusable(tail)) {
      candidates.push(tail)
    }
  }
  if (candidates.length > 0) {
    return candidates.sort((a, b) => b.length - a.length)[0]
  }

  const parts = text.split(/##\s*分析过程|##分析过程/)
  const remainder = parts
    .map((p) => p.trim())
    .filter((p) => p && !looksLikeLiveProcessChunk(p) && !isProcessTemplateDump(p))
    .join('\n\n')
  const cleaned = sanitizeFinalAnswerText(remainder)
  if (cleaned && !isProcessTemplateDump(cleaned) && !isFinalAnswerUnusable(cleaned)) {
    return cleaned
  }
  return ''
}

File: backend/content_filters.py
"""Separate filtering rules for process steps vs final answers.

过程信息（步骤列表、工具中间态）与最终结果（模型报告）使用不同规则，
避免过程侧启发式误伤最终输出。
"""

from __future__ import annotations

import re

from tool_summarize import strip_process_runtime_noise


def has_rich_markdown_report(text: str) -> bool:
    """Substantial markdown report (tables/sections), not a collapsed template."""
    t = (text or "").strip()
    if len(t) < 120:
        return False
    headers = len(re.findall(r"^#{1,6}\s+\S", t, re.MULTILINE))
    has_table = "|" in t and "---" in t
    if headers >= 2 and has_table:
        return True
    if headers >= 1 and has_table and len(t) >= 200:
        return True
    return False


def is_process_template_dump(text: str) -> bool:
    """Collapsed ## 分析过程 / 步骤模板，仅用于过程区判断。"""
    t = (text or "").strip()
    if not t:
        return True
    compact = t.replace(" ", "")
    if "##分析过程" in compact or "###步骤" in compact:
        return True
    if any(
        token in t
        for token in (
            "-类型:",
            "- 类型:",
            "等待执行",
            "等待返回",
            "状态:进行中",
            "状态:等待",
            "-状态:进行中",
            "-状态:等待",
        )
    ):
        return True
    if t.count("✅") >= 2 and not any(
        marker in t for marker in ("QED", "评分", "pocket", "PDB", "|", "对接", "hERG", "BBB")
    ):
        return True
    return False


def is_final_answer_unusable(text: str) -> bool:
    """最终回答是否应丢弃并改用步骤合成兜底（规则比过程更宽松）。"""
    t = (text or "").strip()
    if not t:
        return True
    if has_rich_markdown_report(t):
        return False
    if is_process_template_dump(t):
        return True
    if '"pocket_id":' in t and "|" not in t and len(t) < 200:
        return True
    if '"molecules":' in t and "|" not in t and len(t) < 200:
        return True
    return False


def sanitize_process_step_text(text: str) -> str:
    """过程步骤字段：去除 exec 中间态、运行噪声。"""
    return strip_process_runtime_noise(text or "")


def sanitize_final_answer_text(text: str) -> str:
    """最终回答：仅去掉误入的过程段/告警尾注，不套用过程噪声过滤。"""
    t = (text or "").strip()
    if not t:
        return ""
    for marker in ("## 分析过程", "##分析过程"):
        idx = t.find(marker)
        if idx >= 0:
            t = t[:idx].strip()
    noise = t.find("\n⚠️")
    if noise >= 0:
        t = t[:noise].strip()
    exec_fail = t.find("\n⚠️ 🛠️ Exec failed:")
    if exec_fail >= 0:
        t = t[:exec_fail].strip()
    return t.strip()


# Back-compat alias used by tests / gradual migration
def is_low_quality_final_answer(text: str) -> bool:
    return is_final_answer_unusable(text)


_TRAILING_FINAL_START = re.compile(
    r"(?:"
    r"✅\s*\*\*第\s*\d+\s*步"
    r"|✅\s*\*\*步骤\s*\d+"
    r"|(?:^|\n)##\s*结果摘要"
    r"|(?:^|\n)###\s*🔬"
    r"|(?:^|\n)#\s+🔬"
    r"|(?:^|\n)###\s*🫁"
    r")",
    re.MULTILINE,
)
_PROCESS_HEAD = re.compile(r"##\s*分析过程|##分析过程")


def _looks_like_live_process_chunk(text: str) -> bool:
    t = (text or "").strip()
    if not t:
        return True
    if "### 步骤" in t and "- 类型:" in t and "- 状态:" in t:
        return True
    if "- 工具数:" in t and "### 步骤" in t:
        return True
    return False


def _final_answer_quality_score(text: str) -> int:
    t = (text or "").strip()
    if not t:
        return 0
    score = len(t)
    if has_rich_markdown_report(t):
        score += 10_000
    # 步骤合成兜底（**技能名** 开头短列表）权重低于模型报告
    if re.match(r"^\*\*[^*]+\*\*\s*\n\s*[-|]", t) and len(t) < 400:
        score -= 2_000
    return score


def extract_trailing_model_final(raw: str) -> str:
    """Model prose after embedded process blocks when ## 最终回答 is missing."""
    text = (raw or "").strip()
    if not text:
        return ""

    candidates: list[str] = []
    for m in _TRAILING_FINAL_START.finditer(text):
        tail = sanitize_final_answer_text(text[m.start() :])
        if tail and not is_process_template_dump(tail) and not is_final_answer_unusable(tail):
            candidates.append(tail)
    if candidates:
        return max(candidates, key=len)

    remainder_parts: list[str] = []
    for seg in _PROCESS_HEAD.split(text):
        part = seg.strip()
        if not part or _looks_like_live_process_chunk(part) or is_process_template_dump(part):
            continue
        remainder_parts.append(part)
    remainder = sanitize_final_answer_text("\n\n".join(remainder_parts))
    if remainder and not is_process_template_dump(remainder) and not is_final_answer_unusable(remainder):
        return remainder
    return ""
